## Example for generators using many values

In [28]:
import random
from datetime import datetime 
# import mem_profile

### First, using Lists

In [14]:
namesList = ['Max', 'Ella', 'Monica', 'Mat', 'Amy']
majorList = ['Engg', 'CS', 'Arts', 'Hist', 'Med']

# def peepList(noPeep) -> str:
def peepList(noPeep):
    result = []
    for i in range(noPeep):
        person = {
            'ID':i,
            'name':random.choice(namesList),
            'major':random.choice(majorList)
        }
        result.append(person)
    return result
    # return person.to_string()

listPeep = peepList(10)
print(listPeep)

[{'ID': 0, 'name': 'Monica', 'major': 'Arts'}, {'ID': 1, 'name': 'Monica', 'major': 'Med'}, {'ID': 2, 'name': 'Monica', 'major': 'Engg'}, {'ID': 3, 'name': 'Ella', 'major': 'Med'}, {'ID': 4, 'name': 'Max', 'major': 'Med'}, {'ID': 5, 'name': 'Mat', 'major': 'Engg'}, {'ID': 6, 'name': 'Monica', 'major': 'Engg'}, {'ID': 7, 'name': 'Mat', 'major': 'Engg'}, {'ID': 8, 'name': 'Ella', 'major': 'Engg'}, {'ID': 9, 'name': 'Ella', 'major': 'Arts'}]


In [35]:
namesList = ['Max', 'Ella', 'Monica', 'Mat', 'Amy']
majorList = ['Engg', 'CS', 'Arts', 'Hist', 'Med']

In [ ]:
def peepListAdv(noPeep) -> str:
# def peepList(noPeep):
    # result = []
    for i in range(noPeep):
        person = {
            'ID':i,
            'name':random.choice(namesList),
            'major':random.choice(majorList)
        }
        # result.append(person)
    # return result
    return str(person)

listPeep2 = peepList(10)
print(listPeep2)

{'ID': 9, 'name': 'Amy', 'major': 'Hist'}


In [30]:
t1 = datetime.now()
listPeep3 = peepList(1000000)
t2 = datetime.now()

print(f"Time taken by List: {t2-t1}")

Time taken by List: 0:00:00.495463


In [33]:
t1 = datetime.now()
listPeep3 = peepListAdv(1000000)
t2 = datetime.now()

print(f"Time taken by List: {t2-t1}")

Time taken by List: 0:00:00.497942


## Second, using generators

In [36]:
def genPeep(noPeep) -> str:
    
    for i in range(noPeep):
        person = {
            'ID':i,
            'Name':random.choice(namesList),
            'Major':random.choice(majorList)
        }
        yield person

In [39]:
t1 = datetime.now()
listPeep4 = genPeep(1000000)
t2 = datetime.now()

print(f"Time taken by List: {t2-t1}")

Time taken by List: 0:00:00.000049


## Comparing:
    List: 0:00:00.497942
    Gen.: 0.:00:00.000049

Keep in mind that the 'gen' haven't been generated yet.